src/libinjector/linux/methods/linux_shellcode.c

```diff
@@ -0,0 +1,375 @@
+
+    return event;
+}
+
+static event_response_t cleanup(drakvuf_t drakvuf, drakvuf_trap_info_t* info, bool clear_trap)
+{
+    fprintf(stderr, "Doing premature cleanup\n");
+    injector_t injector = (injector_t)info->trap->data;
+
+    // restore regs
+    memcpy(info->regs, &injector->saved_regs, sizeof(x86_registers_t));
+
+    if (clear_trap)
+        free_bp_trap(drakvuf, injector, info->trap);
+
+    // since we are jumping to some arbitrary step, we will set this
+    injector->step_override = true;
+
+    // give the last step
+    injector->step = STEP5;
+
+    return VMI_EVENT_RESPONSE_SET_REGISTERS;
+}
+
+static bool setup_mmap_trap(drakvuf_t drakvuf, drakvuf_trap_info_t* info)
+{
+    injector_t injector = info->trap->data;
+
+    injector->bp = g_try_malloc0(sizeof(drakvuf_trap_t));
+
+    injector->bp->type = BREAKPOINT;
+    injector->bp->name = "injector_mmap_trap";
+    injector->bp->cb = injector_int3_userspace_cb;
+    injector->bp->data = injector;
+    injector->bp->breakpoint.lookup_type = LOOKUP_DTB;
+    injector->bp->breakpoint.dtb = info->regs->cr3;
+    injector->bp->breakpoint.addr_type = ADDR_VA;
+    injector->bp->breakpoint.addr = injector->virtual_memory_addr;
+    injector->bp->ttl = UNLIMITED_TTL;
+    injector->bp->ah_cb = NULL;
+
+    if ( drakvuf_add_trap(drakvuf, injector->bp) )
+    {
+        PRINT_DEBUG("mmap trap successful\n");
+        return true;
+    }
+    else
+    {
+        fprintf(stderr, "Couldn't trap mmap location\n");
+        return false;
+    }
+}
+
+static bool write_shellcode_to_mmap_location(drakvuf_t drakvuf, drakvuf_trap_info_t* info)
+{
+    injector_t injector = (injector_t)info->trap->data;
+
+    ACCESS_CONTEXT(ctx,
+        .translate_mechanism = VMI_TM_PROCESS_DTB,
+        .dtb = info->regs->cr3,
+        .addr = injector->virtual_memory_addr
+    );
+
+    size_t bytes_write = 0;
+
+    vmi_instance_t vmi = drakvuf_lock_and_get_vmi(drakvuf);
+
+    if ( vmi_write(vmi, &ctx, injector->buffer.len, injector->buffer.data, &bytes_write) != VMI_SUCCESS )
+    {
+        drakvuf_release_vmi(drakvuf);
+        fprintf(stderr, "Could not write the shellcode in memory\n");
+        return false;
+    }
+
+    PRINT_DEBUG("Shellcode write success in memory\n");
+    print_hex(injector->buffer.data, injector->buffer.len, bytes_write);
+
+    drakvuf_release_vmi(drakvuf);
+
+    return true;
+}
+
+bool load_shellcode_from_file(injector_t injector, const char* file)
+{
+    FILE* fp = fopen(file, "rb");
+    if (!fp)
+    {
+        fprintf(stderr, "Shellcode file (%s) not existing\n", file);
+        return false;
+    }
+
+    fseek(fp, 0, SEEK_END);
+    if ( (injector->buffer.len = ftell (fp)) < 0 )
+    {
+        fprintf(stderr, "ftell returned -1\n");
+        fclose(fp);
+        return false;
+    }
+    rewind(fp);
+
+    // we are adding +1 as we will append ret instruction for restoring the state of the VM
+    injector->buffer.data = g_try_malloc0(injector->buffer.len + 1);
```

Comment:
```suggestion
    injector->buffer.data = g_malloc0(injector->buffer.len + 1);
```

Reply:
If i'll be doing this, then do I remove the proceeding `if (!injector->buffer.data)`?

Reply:
Yes
![image](https://user-images.githubusercontent.com/26001306/128817903-bbbb88d5-b67e-41e9-9b64-7a73a56375f7.png)

Reply:
ooh.

btw, are the programs exited cleanly when it aborts?

Reply:
Done